### PR TITLE
Tx verifier homomorphic tests

### PR DESCRIPTION
**Pull request by azarovh**
Use `TransactionVerificationStrategy` to test homomorphic property of TransactionVerifiers.

Fix merging of values in `TxIndexCache` and block undo cache.

### CODE REVIEW

File: chainstate/tx-verifier/src/transaction_verifier/mod.rs [hunk 1]
```diff
@@ -151,8 +154,7 @@ impl<'a, S: TransactionVerifierStorageRef> TransactionVerifier<'a, S> {
                 let issuance_token_id_getter =
                     || -> Result<Option<TokenId>, ConnectTransactionError> {
                         // issuance transactions are unique, so we use them to get the token id
-                        self.storage_ref
-                            .get_token_id_from_issuance_tx(tx_id)
+                        self.get_token_id_from_issuance_tx(tx_id)
```
Comment (azarovh) [inline chainstate/tx-verifier/src/transaction_verifier/mod.rs:157]: this change makes it possible to issue token in one tx and transfer in another tx in the same block

File: chainstate/src/detail/tx_verification_strategy/default_strategy.rs [hunk 2]
```diff
@@ -81,9 +41,8 @@ impl Default for DefaultTransactionVerificationStrategy {
 }
 
 impl TransactionVerificationStrategy for DefaultTransactionVerificationStrategy {
-    fn connect_block<'a, H, S, M>(
+    fn connect_block<'a, H, S>(
         &self,
-        tx_verifier_maker: M,
```
Comment (TheQuantumPhysicist) [inline chainstate/src/detail/tx_verification_strategy/default_strategy.rs:86]: Why did you remove this?

File: chainstate/src/detail/tx_verification_strategy/disposable_strategy.rs [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+                let fee = tx_verifier
+                    .connect_transactable(
+                        block_index,
+                        BlockTransactableRef::Transaction(block, tx_num),
+                        &block_index.block_height(),
+                        &median_time_past,
+                    )
+                    .map_err(BlockError::StateUpdateFailed)
+                    .log_err()?;
+                let consumed_cache = tx_verifier.consume()?;
```
Comment (TheQuantumPhysicist) [inline chainstate/src/detail/tx_verification_strategy/disposable_strategy.rs:91]: I see this, and all I can think of is that we can write a test that tests every possible combination and asserts that the result is the same. Then, we make that monstrosity part of all framework tests using rstest.

File: chainstate/tx-verifier/src/transaction_verifier/hierarchy.rs [hunk 4]
```diff
@@ -178,18 +168,18 @@ impl<'a, S: TransactionVerifierStorageMut> TransactionVerifierStorageMut
...
-                    undo: undo.clone(),
-                    is_fresh: true,
-                });
-                Ok(())
-            }
-        }
+        self.utxo_block_undo
+            .entry(id)
+            .and_modify(|cached_undo| {
+                cached_undo.undo.append(new_undo.clone());
```
Comment (azarovh) [inline chainstate/tx-verifier/src/transaction_verifier/hierarchy.rs:176]: please pay attention to this line. I have doubts whether it is ok to call `append` to merge `BlockUndo`, because the function is called `set_undo_data` and there is no guarantee that info in `new_undo` actually corresponds to txs with later indexes.  Though this works for our code. What do you think?

File: chainstate/src/detail/tx_verification_strategy/default_strategy.rs (hunk 2 above)
Comment (azarovh) [reply chainstate/src/detail/tx_verification_strategy/default_strategy.rs:86]: To implement disposable strategy I need to create different verifiers that could be modified as well as not. The function limits this to read only `S`.

Comment (azarovh) [reply chainstate/src/detail/tx_verification_strategy/default_strategy.rs:86]: No, I'm wrong. It was the case but it was fixed by changes in chainstate/tx-verifier/src/transaction_verifier/hierarchy.rs lline 108.
Will revert.

File: chainstate/test-framework/src/framework_builder.rs [hunk 5]
```diff
@@ -35,6 +43,7 @@ pub struct TestFrameworkBuilder {
     chainstate_storage: Store,
     custom_orphan_error_hook: Option<Arc<OrphanErrorHandler>>,
     time_getter: TimeGetter,
+    tx_verification_strategy: TxVerificationStrategy,
```
Comment (TheQuantumPhysicist) [inline chainstate/test-framework/src/framework_builder.rs:46]: I don't understand the need to make this an enum. Why not just put the object in here then use it in Chainstate, instead of using this enum as a proxy to match on when building?

Comment (azarovh) [reply chainstate/test-framework/src/framework_builder.rs:46]: What object exactly? It depends on a strategy and I didn't want to make `TestFrameworkBuilder` generic over it.  And `TransactionVerificationStrategy` trait cannot be used for dynamic polymorphism.

Comment (azarovh) [reply chainstate/test-framework/src/framework_builder.rs:46]: Strategy customisation is rare, but if I make it generic then every user of `TestFrameworkBuilder` would have to explicitly specify it. With enums it looks odd, but it is hidden inside the implementation.

File: chainstate/test-suite/src/tests/homomorphism.rs [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+            .unwrap();
+
+        assert_eq!(storage1.dump_raw(), storage2.dump_raw());
+    });
+}
+
+#[rstest]
+#[trace]
+#[case(Seed::from_entropy())]
+fn tokens_homomorphism(#[case] seed: Seed) {
```
Comment (TheQuantumPhysicist) [inline chainstate/test-suite/src/tests/homomorphism.rs:118]: I wonder whether we need more variations... at least throw in TODOs for that if you don't want to do that right now. I think all output types must go in here in a way, and I'm thinking this needs more thought on whether this should be done manually or in some generic form that guarantees that all new types are covered.

Comment (TheQuantumPhysicist) [reply chainstate/test-suite/src/tests/homomorphism.rs:118]: I'm not talking about tokens, specifically, btw.

Comment (TheQuantumPhysicist) [reply chainstate/test-suite/src/tests/homomorphism.rs:118]: Perhaps a connection strategy that does the merges randomly solves this problem. The one I mentioned before in my comment in disposable_strategy

File: chainstate/test-framework/src/framework_builder.rs (hunk 5 above)
Comment (TheQuantumPhysicist) [reply chainstate/test-framework/src/framework_builder.rs:46]: I see the convenience there, but that match... I bet there's a way to make it go away or abstract it away. I don't know, ignore it for now.

File: chainstate/tx-verifier/src/transaction_verifier/mod.rs (hunk 1 above)
Comment (TheQuantumPhysicist) [reply chainstate/tx-verifier/src/transaction_verifier/mod.rs:157]: Seems correct to me.

**Review by stanislav-tkach — APPROVED** [review]
